File: app/src/main/java/com/example/memeshare/MainActivity.java
package com.example.memeshare;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;


import org.json.JSONException;
import org.json.JSONObject;



public class MainActivity extends AppCompatActivity {
    ImageView memeImageView;
    ProgressBar progressBar;
    String memeImageUrl;
    Button shareButton;
    Button nextButton;

    String nsfwValue;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        ActionBar actionBar;
        actionBar = getSupportActionBar();
        ColorDrawable colorDrawable = new ColorDrawable(Color.parseColor("#BD0101"));   // hexadecimal code se lene ke liye
        assert actionBar != null;
        actionBar.setBackgroundDrawable(colorDrawable);

        memeImageView = findViewById(R.id.memeImageView);
        progressBar = findViewById(R.id.progressBar);
        shareButton = findViewById(R.id.shareButton);
        nextButton = findViewById(R.id.nextButton);

        loadMeme();


        shareButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent();
                intent.setAction(Intent.ACTION_SEND);
                intent.setType("text/plain");
                intent.putExtra(Intent.EXTRA_TEXT, "Hey Checkout this Meme\n" + memeImageUrl);
                startActivity((Intent.createChooser(intent,"Share this meme using.....")));
            }
        });

        nextButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                progressBar.setVisibility(View.VISIBLE);

                loadMeme();
            }
        });

    }


    private void checkInternetConnection(){
        ConnectivityManager connMgr = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();       //current active info.


        if(networkInfo != null && networkInfo.isConnected()){

            loadMeme();
        }
        else{
            progressBar.setVisibility(View.GONE);
            memeImageView.setVisibility(View.INVISIBLE);
            shareButton.setEnabled(false);

            Toast.makeText(MainActivity.this, "No Internet Connection", Toast.LENGTH_SHORT).show();
        }
    }

    public void loadMeme(){
        String MEME_URL =  "https://meme-api.com/gimme";


        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.GET, MEME_URL,null, response -> {
            try {
                memeImageView.setVisibility(View.VISIBLE);
                shareButton.setEnabled(true);
                memeImageUrl = response.getString("url");


                nsfwValue = response.getString("nsfw");
                Log.d("URL", memeImageUrl);

                Glide.with(MainActivity.this).load(memeImageUrl).listener(new RequestListener<Drawable>() {// request listner is to get notified for image loading process tha how much it happend
                    @Override
                    public boolean onLoadFailed(@Nullable @org.jetbrains.annotations.Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                        progressBar.setVisibility(View.GONE);
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                        progressBar.setVisibility(View.GONE);
                            if (nsfwValue.equals("false")){

                            }
                        return false;
                    }
                }).into(memeImageView);
            } catch (JSONException e) {

                Log.d("Error", "Something Went Wrong While Fetching Data");
                shareButton.setEnabled(false);
                Toast.makeText(MainActivity.this, "Something Went Wrong While Fetching Data", Toast.LENGTH_SHORT).show();
            }
        }, error -> {
            checkInternetConnection();
        });

        // Add a requestto my RequestQueue.
        MySingleton.getInstance(this).addToRequestQueue(jsonObjectRequest);


    }

}

File: app/src/main/java/com/example/memeshare/MainActivity.kt
package com.example.memeshare

import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target


class MainActivity : AppCompatActivity() {

    var currentImageUrl: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        loadMeme()
    }

    private fun loadMeme(){

        var progressBar = findViewById<ProgressBar>(R.id.progressBar)
        progressBar.visibility = View.VISIBLE

        var memeImageview = findViewById<ImageView>(R.id.memeImageView)





        val queue = Volley.newRequestQueue(this)
         currentImageUrl = "https://meme-api.herokuapp.com/gimme"
       // currentImageUrl = "https://steppschuh-json-porn-v1.p.rapidapi.com/porn/?offset=0&pornId=5073292679446528&producerId=4554967436230656&pornType=4&count=5&actorId=5681034041491456&genreId=5245132710346752"

        // Request a string response from the provided URL.
        val jsonObjectRequest = JsonObjectRequest(Request.Method.GET,currentImageUrl,null,
            { response ->
                currentImageUrl = response.getString("url")
                Glide.with(this).load(currentImageUrl).listener(object : RequestListener<Drawable> {
                    override fun onLoadFailed(
                        e: GlideException?,
                        model: Any?,
                        target: Target<Drawable>?,
                        isFirstResource: Boolean
                    ): Boolean {
                        progressBar.visibility = View.GONE
                        return false
                    }

                    override fun onResourceReady(
                        resource: Drawable?,
                        model: Any?,
                        target: Target<Drawable>?,
                        dataSource: DataSource?,
                        isFirstResource: Boolean
                    ): Boolean {
                        progressBar.visibility = View.GONE
                        return false
                    }
                }).into(memeImageview)

            },


            {Response.ErrorListener {
                Toast.makeText(this,"Something Went Wrong", Toast.LENGTH_LONG).show()
            }

            })

            queue.add(jsonObjectRequest)
    }
    fun shareMeme(view: View) {
        val intent = Intent(Intent.ACTION_SEND)
        intent.putExtra(Intent.EXTRA_TEXT,"Hey chekout this meme $currentImageUrl")
        intent.type= "text/plain"
        val chooser= Intent.createChooser(intent,  "Share this meme using ")
        startActivity(intent)
    }


    fun nextMeme(view: android.view.View) {
        loadMeme()
    }
}

























